core/app/models/spree/payment.rb

```diff
@@ -280,5 +364,20 @@ def create_eligible_credit_event
         })
       end
     end
+
+    def store_state_change(previous_state, new_state)
+      state_changes.create!(
+        previous_state: previous_state,
+        next_state:     new_state,
+        name:           'payment'
+      )
+    end
+
+    def change_state!(new_state)
+      previous_state = state
+      return if new_state == previous_state
```

Comment:
Can we try to make use of `ActiveModel::Dirty` here?

http://api.rubyonrails.org/classes/ActiveModel/Dirty.html